language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -240,34 +240,26 @@ fn parse_copyable_val_<'input>(
     Ok(spanned(start_loc, end_loc, val))
 }
 
-// Exp = BinopExp = CmpOpExp
-// CmpOpExp = (OrExp CmpOp OrExp) || OrExp
-// OrExp = (AndExp OrOp AndExp) || AndExp
-// AndExp = (XorExp AndOp XorExp) || XorExp
-// XorExp = (BinOrExp XorOp BinOrExp) || BinOrExp
-// BinOrExp = (BinAndExp BinOrOp BinAndExp) || BinAndExp
-// BinAndExp = (AddSubExp BinAndOp AddSubExp) || AddSubExp
-// AddSubExp = (FactorExp AddSubOp FactorExp) || FactorExp
-// FactorExp = (UnaryExp FactorOp UnaryExp) || UnaryExp
-
+// Get the precedence of a binary operator.
 fn get_precedence(token: &Tok) -> u32 {
     match token {
-        Tok::EqualEqual => 1,
-        Tok::ExclaimEqual => 1,
-        Tok::Less => 1,
-        Tok::Greater => 1,
-        Tok::LessEqual => 1,
-        Tok::GreaterEqual => 1,
+        // Reserved minimum precedence value is 1 (specified in parse_exp_)
         Tok::PipePipe => 2,
         Tok::AmpAmp => 3,
-        Tok::Caret => 4,
+        Tok::EqualEqual => 4,
+        Tok::ExclaimEqual => 4,
+        Tok::Less => 4,
+        Tok::Greater => 4,
+        Tok::LessEqual => 4,
+        Tok::GreaterEqual => 4,
         Tok::Pipe => 5,
-        Tok::Amp => 6,
-        Tok::Plus => 7,
-        Tok::Minus => 7,
-        Tok::Star => 8,
-        Tok::Slash => 8,
-        Tok::Percent => 8,
+        Tok::Caret => 6,
+        Tok::Amp => 7,
+        Tok::Plus => 8,
+        Tok::Minus => 8,
+        Tok::Star => 9,
+        Tok::Slash => 9,
+        Tok::Percent => 9,
         _ => 0, // anything else is not a binary operator
```

Comment:
Why 0 here? should we make it an option?

Reply:
Making it an option seems like a good idea. I will give that a try.

Reply:
OK, I tried it. I don't like the result. The way that parse_rhs_of_binary_exp() works is to continue parsing binary operators as long as they have at least the specified minimum precision. It starts off with a minimum precision of 1, so returning 0 for the precedence of other tokens is a natural way to stop that parsing. Using an Option type would mean that the very simple check for "precedence >= min_prec" would need to first check for None and then check the Some values. This happens twice in the loop in parse_rhs_of_binary_exp so it would clutter up the code and make it harder to follow without (IMO) making it more clear in any real way. I will try adding a comment to explain this better.